### PR TITLE
terraform-plugin-docs: init at v0.16.0

### PR DESCRIPTION
**Pull request by catouc**
## Description of changes

Add [terraform-plugin-docs](https://github.com/hashicorp/terraform-plugin-docs/) at version `v0.16.0`

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/misc/terraform-plugin-docs/default.nix
Comment (anthonyroussel) [inline pkgs/development/tools/misc/terraform-plugin-docs/default.nix:15]:
```suggestion
    hash = "sha256-5vbi69GMgkzvN3aEQbNTbk99rg+kfvAvUrdDsuyIm9s=";
  };

  subPackages = [ "cmd/tfplugindocs" ];
  vendorHash = "sha256-AjW6BokLVDkIWXToJ7wNq/g19xKTAfpQ/gVlKCV5qw0=";
```

I'm not sure, but I think `sha256` / `vendorSha256` are deprecated in favor of `hash` / `vendorHash`.

File: pkgs/development/tools/misc/terraform-plugin-docs/default.nix
Comment (anthonyroussel) [inline pkgs/development/tools/misc/terraform-plugin-docs/default.nix:16]: What do you think about passing the version number in the ldlfags:

```nix
  ldflags = [ "-s" "-w" "-X main.version=${version}" "-X main.commit=${src.rev}" ];
```

so that tfplugindocs returns the correct version number:

```console
$ tfplugindocs --version
tfplugindocs Version 0.16.0 from commit v0.16.0
```

instead of:

```console
$ tfplugindocs --version
tfplugindocs Version dev
````

?

**Review by anthonyroussel — COMMENTED** [review]
Result of `nixpkgs-review pr 249387` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>terraform-plugin-docs</li>
  </ul>
</details>

**Review by anthonyroussel — APPROVED** [review]
LGTM :heavy_check_mark: 

`tfplugindocs generate` works when compiling the `terraform-provider-random` Terraform plugin.

It only works if `go` and `terraform` binaries are already in the path.

I'm not sure, but maybe we should consider:

* adding them as runtime dependencies in the `propagatedBuildInputs` list
* use `wrapProgram` to ensure these binaries are in the `PATH`
 
What do you think?

**Comment by catouc** [comment]
> I'm not sure, but maybe we should consider:
> 
> * adding them as runtime dependencies in the `propagatedBuildInputs` list
> * use `wrapProgram` to ensure these binaries are in the `PATH`

Yeah that sounds like a good idea! Should have tested it without my local env 🙃
Will play around with that and update, it's a little weird that it needs `go` in the path tbh but maybe they just call out to a process.

**Comment by catouc** [comment]
Hmm I made a lot of the patches work, but it seem that something is probably ending in a circular dependency on the postInstall wrapper:

```
error: output '/nix/store/m6cwwqfnvw3z1prxzhlfp3f9bpgjx9ky-terraform-plugin-docs-0.16.0' is not allowed to refer to the following paths:
         /nix/store/l9n531fcx4grk3wvfbigya3p66scj61h-go-1.20.6
```

@anthonyroussel Any ideas on if I may have missed something? I purposely left the commits separate, once we have a working setup I'd squash the history together but for debugging it's probably nicer to have commits separate.

**Comment by anthonyroussel** [comment]
> Hmm I made a lot of the patches work, but it seem that something is probably ending in a circular dependency on the postInstall wrapper:
> 
> ```
> error: output '/nix/store/m6cwwqfnvw3z1prxzhlfp3f9bpgjx9ky-terraform-plugin-docs-0.16.0' is not allowed to refer to the following paths:
>          /nix/store/l9n531fcx4grk3wvfbigya3p66scj61h-go-1.20.6
> ```
> 
> @anthonyroussel Any ideas on if I may have missed something? I purposely left the commits separate, once we have a working setup I'd squash the history together but for debugging it's probably nicer to have commits separate.

@catouc Hello

I think you need to set `allowGoReference` to true because tfplugindocs is coupled to the go compiler.

See https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/go/module.nix#L33

I think something like that should work:

```nix
{ lib, buildGoModule, fetchFromGitHub, makeWrapper, go, terraform }:

{
  # ...

  allowGoReference = true;

  nativeBuildInputs = [ makeWrapper ];

  postInstall = ''
    wrapProgram $out/bin/tfplugindocs \
      --prefix PATH : ${lib.makeBinPath [ go terraform ]}
  '';

  # ...
}
```

What do you think?

**Comment by catouc** [comment]
Ah that does it, tested it - works with an unset `PATH` variable so I'm thinking we might have it, thanks for all of your help in this @anthonyroussel ! ❤️

**Review by anthonyroussel — APPROVED** [review]
@catouc You're welcome :)

I've tested tfplugindocs with the terraform-provider-random plugin, and it works perfectly!

---
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

Result of `nixpkgs-review pr 249387` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>terraform-plugin-docs</li>
</ul>
</details>

**Comment by catouc** [comment]
Got around to rebasing finally. @anthonyroussel any further thing we need to get this merged? Sorry work got a little busy around me (:

**Review by anthonyroussel — APPROVED** [review]
Result of `nixpkgs-review pr 249387` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>terraform-plugin-docs</li>
</ul>
</details>

File: pkgs/development/tools/misc/terraform-plugin-docs/default.nix
Comment (anthonyroussel) [inline pkgs/development/tools/misc/terraform-plugin-docs/default.nix:36]:
```suggestion
    mainProgram = "tfplugindocs";
    maintainers = with maintainers; [ catouc ];
```

File: pkgs/development/tools/misc/terraform-plugin-docs/default.nix
Comment (anthonyroussel) [inline pkgs/development/tools/misc/terraform-plugin-docs/default.nix:34]:
```suggestion
    changelog = "https://github.com/hashicorp/terraform-plugin-docs/releases/tag/v${version}";
    homepage = "https://github.com/hashicorp/terraform-plugin-docs";
```

File: pkgs/development/tools/misc/terraform-plugin-docs/default.nix
Comment (anthonyroussel) [inline pkgs/development/tools/misc/terraform-plugin-docs/default.nix:36]:
```suggestion
    maintainers = with maintainers; [ catouc ];
    platforms = platforms.linux ++ platforms.darwin;
```

You can also enable darwin support for this package if you want.
Looks like terraform-plugin-docs upstream supports darwin.

https://github.com/hashicorp/terraform-plugin-docs/blob/main/.goreleaser.yml#L14

File: pkgs/development/tools/misc/terraform-plugin-docs/default.nix
Comment (anthonyroussel) [inline pkgs/development/tools/misc/terraform-plugin-docs/default.nix:27]: Why did you add musl in nativeBuildInputs and ldflags?

File: pkgs/development/tools/misc/terraform-plugin-docs/default.nix
Comment (anthonyroussel) [inline pkgs/development/tools/misc/terraform-plugin-docs/default.nix:32]:
```suggestion
  passthru.tests.version = testers.testVersion {
    package = terraform-plugin-docs;
    command = "tfplugindocs --version";
  };
```

I recommend adding `passthru.tests.version` to test the final binary.

File: pkgs/development/tools/misc/terraform-plugin-docs/default.nix
Comment (catouc) [reply pkgs/development/tools/misc/terraform-plugin-docs/default.nix:27]: This is fairly straight from https://nixos.wiki/wiki/Go to ensure statically compiling works without issues. Seemingly glibc doesn't like static linking as much:

>Linking against glibc.static does not really work because glibc does not really like static linking. You get a warning like warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking. To really create a static build, use musl.

Comment (anthonyroussel) [reply pkgs/development/tools/misc/terraform-plugin-docs/default.nix:27]: @catouc Hello. Thanks for the explanation!

Using musl breaks darwin support, so I think you can rollback the `platforms` line (L44) to:

```nix
    platforms = platforms.linux;
```

What do you think?

**Review by anthonyroussel — APPROVED** [review]
Result of `nixpkgs-review pr 249387` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>terraform-plugin-docs</li>
</ul>
</details>

File: pkgs/development/tools/misc/terraform-plugin-docs/default.nix
Comment (catouc) [reply pkgs/development/tools/misc/terraform-plugin-docs/default.nix:27]: For now I feel that would be the route, with a future to figure out on if we can add glibc back for darwin support. Did push the change for the support :)

**Review by anthonyroussel — APPROVED** [review]
Great work! I think the PR is now ready to be merged :)

Result of `nixpkgs-review pr 249387` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>terraform-plugin-docs</li>
</ul>
</details>

**Comment by anthonyroussel** [comment]
Hello @marsam

I saw in the git log that you are one of the most active maintainer of the terraform package ecosystem on nixpkgs, so you might be the right person who could help us :)

This PR from @catouc to add the terraform-plugin-docs package to nixpkgs has been blocked and waiting for another reviewer and merger for a few weeks.

Would you be ok and available to help us unblock this PR, please?

**Comment by marsam** [comment]
Thanks for the heads up. 

I think it's already packaged as [`tfplugindocs`](https://github.com/NixOS/nixpkgs/blob/6895e18ae034c15e5912a88936b80000bc3c154a/pkgs/development/tools/tfplugindocs/default.nix).  Would you mind moving your changes to that expression?

Also, IMHO `terraform-plugin-docs` is a more adequate name, since it's discoverable and it's used by [AUR](https://repology.org/project/terraform-plugin-docs/versions). 
What do you think of renaming `tfplugindocs` to `terraform-plugin-docs`?

**Comment by anthonyroussel** [comment]
> Thanks for the heads up.
> 
> I think it's already packaged as [`tfplugindocs`](https://github.com/NixOS/nixpkgs/blob/6895e18ae034c15e5912a88936b80000bc3c154a/pkgs/development/tools/tfplugindocs/default.nix). Would you mind moving your changes to that expression?
> 
> Also, IMHO `terraform-plugin-docs` is a more adequate name, since it's discoverable and it's used by [AUR](https://repology.org/project/terraform-plugin-docs/versions). What do you think of renaming `tfplugindocs` to `terraform-plugin-docs`?

@marsam Thank you for your help!
Oops, I didn't realise there was already a package for terraform-plugin-docs 😅
I created another PR to rename the package: https://github.com/NixOS/nixpkgs/pull/264756

**Comment by catouc** [comment]
I suppose with https://github.com/NixOS/nixpkgs/pull/264756 merged we don't need to keep this open.

**Pull request closed without merging**
